File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        int ans  = bfs(root, -101);
        return ans;
    }

    private int bfs(TreeNode currentNode, int maxValue){
        if(currentNode == null)
            return 0;
        int ans = 0;
        if(maxValue <= currentNode.val){
            maxValue = currentNode.val;
            ans++;
        }
        int rightChild = bfs(currentNode.right, maxValue);
        int leftChild = bfs(currentNode.left, maxValue);
        ans = ans + rightChild + leftChild;
        return ans;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.ans = 0

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        leftLen = self.longestPathChild(root.left)
        rightLen = self.longestPathChild(root.right)
        self.ans = max (self.ans, rightLen + leftLen)
        return self.ans

    def longestPathChild(self, node: Optional[TreeNode]) ->int :
        if not node:
            return 0

        leftLen = self.longestPathChild(node.left)
        rightLen = self.longestPathChild(node.right)
        
        self.ans = max(self.ans, rightLen + leftLen)

        return max(leftLen, rightLen) + 1


